algorand/portal_core.py

```diff
@@ -0,0 +1,550 @@
+
+                # Now, lets go grab the raw byte
+                byte_offset.store((sequence.load() / Int(8)) % Int(max_bytes)),
+                b.store(blob.get_byte(Int(1), byte_offset.load())),
+
+                # I would hope we've never seen this packet before...   throw an exception if we have
+                MagicAssert(GetBit(b.load(), sequence.load() % Int(8)) == Int(0)),
+
+                # Lets mark this bit so that we never see it again
+                blob.set_byte(Int(1), byte_offset.load(), SetBit(b.load(), sequence.load() % Int(8), Int(1))),
+
+                blob.meta(Int(1), Bytes("duplicate"))
+            )
+
+        STATELESS_LOGIC_HASH = App.globalGet(Bytes("vphash"))
+
+        def verifyVAA():
+            i = ScratchVar()
+            a = ScratchVar()
+            total_guardians = ScratchVar()
+            guardian_keys = ScratchVar()
+            num_sigs = ScratchVar()
+            off = ScratchVar()
+            digest = ScratchVar()
+            hits = ScratchVar()
+            s = ScratchVar()
+            eoff = ScratchVar()
+            guardian = ScratchVar()
+
+            return Seq([
+                # We have a guardian set?  We have OUR guardian set?
+                MagicAssert(Txn.accounts[2] == get_sig_address(Btoi(Extract(Txn.application_args[1], Int(1), Int(4))), Bytes("guardian"))),
+                blob.checkMeta(Int(2), Bytes("guardian")),
+                # Lets grab the total keyset
+                total_guardians.store(blob.get_byte(Int(2), Int(0))),
+                guardian_keys.store(blob.read(Int(2), Int(1), Int(1) + Int(20) * total_guardians.load())),
+
+                # I wonder if this is an expired guardian set
+                s.store(Btoi(blob.read(Int(2), Int(1000), Int(1008)))),
+                If(s.load() != Int(0),
+                   MagicAssert(Txn.first_valid() < s.load())),
+
+                hits.store(Bytes("base16", "0x00000000")),
+
+                # How many signatures are in this vaa?
+                num_sigs.store(Btoi(Extract(Txn.application_args[1], Int(5), Int(1)))),
+
+                # Lets create a digest of THIS vaa...
+                off.store(Int(6) + (num_sigs.load() * Int(66))),
+                digest.store(Keccak256(Keccak256(Extract(Txn.application_args[1], off.load(), Len(Txn.application_args[1]) - off.load())))),
+
+                # We have enough signatures?
+                MagicAssert(And(
+                    total_guardians.load() > Int(0),
+                    num_sigs.load() <= total_guardians.load(),
+                    num_sigs.load() > ((total_guardians.load() * Int(2)) / Int(3)),
+                    )),
+
+
+                # Point it at the start of the signatures in the VAA
+                off.store(Int(6)),
+
+                For(
+                        i.store(Int(0)),
+                        i.load() <= Txn.group_index(),
```

Comment:
Why `<=` here? Down below we have to now check
```
                                [a.load() == Bytes("verifyVAA"), Seq([])],
```
which is the `==` case. Might as well do `<` and remove that case? I don't think it's possible to have
```
verifySigs
verifySigs
verifyVAA
verifySigs
...
verifyVAA
```

Reply:
I want the same asserts executed on verifyVAA on lines 404-408 as I want for all the verifySigs.  If I make it just a <, then I'd have to copy those asserts to another spot to execute them for Txn.  Doing in this way saves code

Reply:
Now as another interesting point, if we were to modify this to not start at 0 but instead to walk backwards from the current index...   then you would be "allowed" to put two wormhole ops into a single txn group so that you could do two operations transactionally...

Reply:
That's sounds good, do you think it's worth switching over to that pattern? It would also mean better composition not just between wormhole txs but other contracts too

Reply:
Yeah, except... 1) We'd have to tell the auditors about the change and they would  not like that we are changing the contract they have been auditing,  2) We would work backwards until we saw a function we didn't understand OR we have consumed all the signatures in the vaa?  lets chat after the meeting